public/components/incontext_insight/generate_popover_body.tsx

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useState } from 'react';
+import { i18n } from '@osd/i18n';
+import {
+  EuiButton,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiIcon,
+  EuiPanel,
+  EuiSpacer,
+  EuiText,
+} from '@elastic/eui';
+import { IncontextInsight as IncontextInsightInput } from '../../types';
+import { getNotifications, IncontextInsightRegistry } from '../../services';
+import { HttpSetup } from '../../../../../src/core/public';
+import { ASSISTANT_API } from '../../../common/constants/llm';
+import { getAssistantRole } from '../../utils/constants';
+
+export const GeneratePopoverBody: React.FC<{
+  incontextInsight: IncontextInsightInput;
+  registry?: IncontextInsightRegistry;
+  httpSetup?: HttpSetup;
+  closePopover: () => void;
+}> = ({ incontextInsight, registry, httpSetup, closePopover }) => {
+  const [isLoading, setIsLoading] = useState(false);
+  const [isLlmResponded, setIsLlmResponded] = useState(false);
+  const [summary, setSummary] = useState('');
+  const [conversationId, setConversationId] = useState('');
+  const toasts = getNotifications().toasts;
+
+  const onChatContinuation = () => {
+    registry?.continueInChat(incontextInsight, conversationId);
+    closePopover();
+  };
+
+  const onGenerateSummary = (summarizationQuestion: string) => {
+    setIsLoading(true);
+    setIsLlmResponded(false);
+    const summarize = async () => {
+      const contextContent = incontextInsight.contextProvider
+        ? await incontextInsight.contextProvider()
+        : '';
+
+      await httpSetup
+        ?.post(ASSISTANT_API.SEND_MESSAGE, {
+          body: JSON.stringify({
+            messages: [],
+            input: {
+              type: 'input',
+              content: summarizationQuestion,
+              contentType: 'text',
+              context: { content: contextContent, dataSourceId: incontextInsight.datasourceId },
+              promptPrefix: getAssistantRole(incontextInsight.key),
+            },
+          }),
+        })
+        .then((response) => {
+          const interactionLength = response.interactions.length;
+          if (interactionLength > 0) {
+            setConversationId(response.interactions[interactionLength - 1].conversation_id);
+          }
+
+          const messageLength = response.messages.length;
+          if (messageLength > 0 && response.messages[messageLength - 1].type === 'output') {
+            setSummary(response.messages[messageLength - 1].content);
+          }
+          setIsLlmResponded(true);
+        })
+        .catch((error) => {
+          toasts.addDanger(
+            i18n.translate('assistantDashboards.incontextInsight.generateSummaryError', {
+              defaultMessage: 'Generate summary error',
+            })
+          );
+          setIsLoading(false);
```

Comment:
Shall we move this `setIsLoading(false)` to the `.finally` section? Seems the `isLoading` will always be `true` if no error throwed.

Reply:
Realized I could simplify the states of those transitions. The summary response and isLlmResponded is overlapped. The initial implementation using isLoading && isLlmResponded to determine whether to show the summary text box. But that's not necessary, checking the content of summary can achieve the same effect of isLlmResponded.

There might be some lifecycle problem of component. After closing the popover, the summary content is also cleared(should be expected). Might be due to the component is not rendered anymore and garbage collected? So I don't handle the clear summary and conversationId in onChatContinuation.

Reply:
I guess the `GeneratePopoverBody` may unmount after `closePopover` called. The `isOpen` is a part of render condition in the `OuiPopover`. The `GeneratePopoverBody` won't be render if not open. Here is the detailed code: https://github.com/opensearch-project/oui/blob/main/src/components/popover/popover.tsx#L748